Hardcoded #define NDEBUG in yacc/defs.h

https://github.com/ocaml/ocaml/blob/43c320501271c031052e2b62ec34e933a0d8f36a/yacc/defs.h#L18-L20

On a Mac (12.6, Apple clang version 14.0.0 (clang-1400.0.29.202)) this causes an error when -DNDEBUG is passed on the command line:

```
In file included from yacc/error.c:20:
yacc/defs.h:19:9: error: 'NDEBUG' macro redefined [-Werror,-Wmacro-redefined]
#define NDEBUG
        ^
<command line>:9:9: note: previous definition is here
#define NDEBUG 1
```

--- a/Makefile
+++ b/Makefile
@@ -1114,6 +1114,9 @@ ocamlyacc_MODULES = $(ocamlyacc_WSTR_MODULE) $(ocamlyacc_OTHER_MODULES)
 
 ocamlyacc_OBJECTS = $(ocamlyacc_MODULES:=.$(O))
 
+# Do not compile assertions in ocamlyacc
+ocamlyacc_CPPFLAGS = -DNDEBUG
+
 .PHONY: ocamlyacc
 ocamlyacc: $(ocamlyacc_PROGRAM)$(EXE)
 
@@ -1124,6 +1127,8 @@ clean::
 	rm -f $(ocamlyacc_MODULES:=.o) $(ocamlyacc_MODULES:=.obj)
 
 $(ocamlyacc_OTHER_MODULES:=.$(O)): yacc/defs.h
+
+$(ocamlyacc_OTHER_MODULES:=.$(O)): OC_CPPFLAGS += $(ocamlyacc_CPPFLAGS)
 
 # The Menhir-generated parser
 

--- a/otherlibs/systhreads/st_pthreads.h
+++ b/otherlibs/systhreads/st_pthreads.h
@@ -15,7 +15,6 @@
 
 /* POSIX thread implementation of the "st" interface */
 
-#include <assert.h>
 #include <errno.h>
 #include <string.h>
 #include <stdio.h>

--- a/yacc/defs.h
+++ b/yacc/defs.h
@@ -15,10 +15,6 @@
 
 /* Based on public-domain code from Berkeley Yacc */
 
-#if !defined(DEBUG) && !defined(NDEBUG)
-#define NDEBUG
-#endif
-
 #include <assert.h>
 #include <ctype.h>
 #include <errno.h>

--- a/yacc/defs.h
+++ b/yacc/defs.h
@@ -15,7 +15,7 @@
 
 /* Based on public-domain code from Berkeley Yacc */
 
-#ifndef DEBUG
+#if !defined(DEBUG) && !defined(NDEBUG)
 #define NDEBUG
 #endif
 

